extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -0,0 +1,417 @@
+      usableForEncryption: ! await Catch.doesReject(pubkey.getEncryptionKey()),
+      usableButExpired: await OpenPGPKey.usableButExpired(pubkey, exp, expired),
+      usableForSigning: await Catch.doesReject(pubkey.getSigningKey()),
+      emails,
+      // tslint:disable-next-line: no-unsafe-any
+      identities: pubkey.users.map(u => u.userId).filter(u => !!u && u.userid && Str.parseEmail(u.userid).email).map(u => u!.userid).filter(Boolean) as string[],
+      lastModified,
+      expiration: exp instanceof Date ? exp : undefined,
+      created: pubkey.getCreationTime(),
+      checkPassword: _text => Promise.resolve(false),
+      fullyDecrypted: pubkey.isPublic() ? true /* public keys are always decrypted */ : pubkey.isFullyDecrypted(),
+      fullyEncrypted: pubkey.isPublic() ? false /* public keys are never encrypted */ : pubkey.isFullyEncrypted(),
+      isPublic: pubkey.isPublic(),
+      isPrivate: pubkey.isPrivate(),
+    } as Key);
+    pkey.checkPassword = async passphrase => PgpKey.decrypt(await OpenPGPKey.parse(OpenPGPKey.armor(pkey)), passphrase);
+    const extensions = pkey as unknown as { raw: string, [internal]: OpenPGP.key.Key };
+    extensions[internal] = pubkey;
+    extensions.raw = raw || pubkey.armor();
+    return pkey;
+  }
+
+  /**
+   * Returns signed data if detached=false, armored
+   * Returns signature if detached=true, armored
+   */
+  public static sign = async (signingPrivate: Key, data: string, detached = false): Promise<string> => {
+    const signingPrv = OpenPGPKey.unwrap(signingPrivate);
+    const message = opgp.cleartext.fromText(data);
+    const signRes = await opgp.sign({ message, armor: true, privateKeys: [signingPrv], detached });
+    if (detached) {
+      if (typeof signRes.signature !== 'string') {
+        throw new Error('signRes.signature unexpectedly not a string when creating detached signature');
+      }
+      return signRes.signature;
+    }
+    return await opgp.stream.readToEnd((signRes as OpenPGP.SignArmorResult).data);
+  }
+
+  public static revoke = async (key: Key): Promise<string | undefined> => {
+    let prv = OpenPGPKey.unwrap(key);
+    if (! await prv.isRevoked()) {
+      prv = await prv.revoke({});
+    }
+    const certificate = await prv.getRevocationCertificate();
+    if (!certificate) {
+      return undefined;
+    } else if (typeof certificate === 'string') {
+      return certificate;
+    } else {
+      return await opgp.stream.readToEnd(certificate);
+    }
+  }
+
+  public static armor = (pubkey: Key): string => {
+    if (pubkey.type !== 'openpgp') {
+      throw new Error('Unsupported key type: ' + pubkey.type);
+    }
+    const extensions = pubkey as unknown as { raw: string };
+    if (!extensions.raw) {
+      throw new Error('Object has type == "openpgp" but no raw key.');
+    }
+    return extensions.raw;
+  }
+
+  private static unwrap = (pubkey: Key) => {
```

Comment:
Maybe wrap/unwrap is indeed a good name

Reply:
I still have some half-baked ideas in my head that I'm not sure if they make sense... for example during refactorings I saw that the Pubkey interface is *mostly* immutable. The only thing that could change it was encrypting and decrypting key. That required some changes in the internal API but I'm thinking whether it would be a good idea to just have separate interfaces for encrypted and decrypted keys and return them from the encrypt/decrypt methods. This is tricky because OpenPGP.js keys are internally mutable so we would either clone the objects or...

The other thing is Pubkeys travelling across realms: they lose the associated OpenPGP.js keys because (obviously) this is not JSON serializable. This doesn't seem to be a big problem now but still something that keeps me awake at night... ;)